### PR TITLE
webpack 4 upgrade

### PR DESCRIPTION
**Pull request by pi0**
**See notes  here #2686**

Checklist:

- [x] Basic required changes
- [x] Timefix is no longer working!
- [x] Fix tests
- [x] Fix windows tests
- [x] Fix font preloading [SKIPED]
- [x] follow best practices of webpack 4 for optimization

Plugins:
- [x] extract-text-webpack-plugin
- [x] html-webpack-plugin
- [x] timefix/warnfix

### CODE REVIEW

File: lib/builder/webpack/base.config.js [hunk 1]
```diff
@@ -28,6 +28,7 @@ module.exports = function webpackBaseConfig({ name, isServer }) {
 
   const config = {
     name,
+    mode: this.options.dev ? 'development' : 'prodoction',
```
Comment (paulgv) [inline lib/builder/webpack/base.config.js]: Small typo in `production` here @pi0

Comment (pi0) [reply lib/builder/webpack/base.config.js]: Fixed thanks.

**Comment by chanlito** [comment]
can't wait for this 😊

**Comment by shellscape** [comment]
FYI, sokra stated that we can't access the `watcher` anymore in `watchRun`:

```
sokra [8:19 AM]
hmm.. I guess you can't access it...
but feel free to send a PR adding a `watching` hooks which allows access to the watching instance
```

**Comment by pi0** [comment]
@shellscape A workaround is coming. We can hijax compilation when calling compiler.watch

**Comment by benfavre** [comment]
I need this so bad. Hope you guys are doing ok on this issue 👍

**Comment by pi0** [comment]
Windows seems is failing tests. Any help is appreciated 🙂

**Comment by clarkdo** [comment]
@pi0 
Current windows tests failure is because of https://github.com/webpack/webpack-dev-middleware/issues/270.
I have tried to apply the https://github.com/webpack/webpack-dev-middleware/pull/276 into my local webpack, it works fine.
I think we can wait the merge and release of webpack-dev-middleware.

**Comment by manniL** [comment]
Webpack dev middleware 3.0.1 has been [released](https://github.com/webpack/webpack-dev-middleware/commit/14de0b99cbdf2d3de54bdc24eeb5ab504d43685f)

**Comment by eljass** [comment]
Do you guys have any ETA for this to be merged?

**Comment by dohomi** [comment]
great news 👍

**Pull request merged**